File: 39-combination-sum/39-combination-sum.cpp
class Solution {
private:
    void findcomb(int first, vector<int> &arr, int target, vector<vector<int>> &ans, vector<int> &ds) {
        if (first == arr.size()) {
            if (target == 0)
                ans.push_back(ds);
            return;
        }
        
        if (arr[first] <= target) {
            ds.push_back(arr[first]);
            findcomb(first, arr, target - arr[first], ans, ds);
            ds.pop_back();
        }
        findcomb(first + 1, arr, target, ans, ds);
    }
    
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> ds;
        findcomb(0, candidates, target, ans, ds);
        return ans;
    }
};

File: 39-combination-sum/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> list = new ArrayList();
        findComb(0, target, candidates, list, new ArrayList());
        return list;
    }
    
    public void findComb(int index, int target, int[] arr, List<List<Integer>> ans, List<Integer> ds) {
        if (index == arr.length) {
            if (target == 0)
                ans.add(new ArrayList(ds));
            return;
        }
        
        if (arr[index] <= target) {
            ds.add(arr[index]);
            findComb(index, target - arr[index], arr, ans, ds);
            ds.remove(ds.size() - 1);
        }
        
        findComb(index + 1, target, arr, ans, ds);
    }
}